pkg/controllermanager/controller/shoot/maintenance/reconciler.go

```diff
@@ -150,12 +162,58 @@ func (r *Reconciler) reconcile(ctx context.Context, log logr.Logger, shoot *gard
 		reasonsForWorkerPoolKubernetesUpdate[w.Name] = reasonForWorkerPoolKubernetesUpdate
 	}
 
-	maintainOperation(shoot)
+	for workerPool, reason := range reasonsForWorkerPoolKubernetesUpdate {
+		if reason != "" {
+			operation := fmt.Sprintf("For worker pool %s: %s", workerPool, reason)
+			operations = append(operations, operation)
+		}
+	}
+
+	operation := maintainOperation(maintainedShoot)
+	if operation != "" {
+		operations = append(operations, "Other task started during maintenance time - "+operation)
+	}
+
+	if operations != nil {
+		patch := client.MergeFrom(shoot.DeepCopy())
+		shoot.Status.LastMaintenance = &gardencorev1beta1.LastMaintenance{
+			Description:   strings.Join(operations, ", "),
+			TriggeredTime: metav1.Time{Time: r.Clock.Now()},
+		}
+
+		// First dry run the update call to check if it can be executed successfully.
+		// If not shoot maintenance is marked as failed and is retried only in
+		// next maintenance window.
+		if err := r.Client.Update(ctx, maintainedShoot, &client.UpdateOptions{
+			DryRun: []string{metav1.DryRunAll},
+		}); err != nil {
+			if hasMaintainNowAnnotation(shoot) {
+				delete(shoot.Annotations, v1beta1constants.GardenerOperation)
+			}
+			shoot.Status.LastMaintenance.Status = pointer.String(fmt.Sprintf("Maintenance failed: %s", err.Error()))
+			if err := r.Client.Status().Patch(ctx, shoot, patch); err != nil {
+				log.Error(err, "Failed to patch shoot status")
+				return err
+			}
+
+			log.Info("Shoot maintenance failed", "error", err)
+			return nil
+		}
+
+		if err := r.Client.Status().Patch(ctx, shoot, patch); err != nil {
+			log.Error(err, "Failed to patch shoot status")
+			return err
+		}
+	}
+
+	// update shoot spec changes in maintenance call
+	shoot.Spec = *maintainedShoot.Spec.DeepCopy()
+	_ = maintainOperation(shoot)
 	maintainTasks(shoot, r.Config)
 
 	// try to maintain shoot, but don't retry on conflict, because a conflict means that we potentially operated on stale
 	// data (e.g. when calculating the updated k8s version), so rather return error and backoff
-	if err := gardenClient.Update(ctx, shoot); err != nil {
+	if err := r.Client.Update(ctx, shoot); err != nil {
```

Comment:
Shouldn't we update the state to `Succeeded` in the `.status.lastMaintenance` now?